src/System.Memory/src/System/Buffers/ReadOnlySequence.cs

```diff
@@ -385,6 +385,34 @@ private ReadOnlySequence<T> SliceImpl(in SequencePosition begin, in SequencePosi
             );
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private SequenceType GetSequenceType()
+        {
+            // We take high order bits of two indexes and move them
+            // to a first and second position to convert to SequenceType
+
+            // if (start < 0  and end < 0)
+            // start >> 31 = -1, end >> 31 = -1
+            // 2 * (-1) + (-1) = -3, result = (SequenceType)3
+
+            // if (start < 0  and end >= 0)
+            // start >> 31 = -1, end >> 31 = 0
+            // 2 * (-1) + 0 = -2, result = (SequenceType)2
+
+            // if (start >= 0  and end >= 0)
+            // start >> 31 = 0, end >> 31 = 0
+            // 2 * 0 + 0 = 0, result = (SequenceType)0
+
+            // if (start >= 0  and end < 0)
+            // start >> 31 = 0, end >> 31 = -1
+            // 2 * 0 + (-1) = -1, result = (SequenceType)1
+
+            return (SequenceType)(-(2 * (_sequenceStart.GetInteger() >> 31) + (_sequenceEnd.GetInteger() >> 31)));
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int GetIndex(in SequencePosition position) => position.GetInteger() & ReadOnlySequence.IndexBitMask;
```

Comment:
You got 2 potential places you could use it in: GetTypeAndIndices (function below)

Reply:
Actually I think it would make sense for this function to take int/uint directly and then do that. No reason you need to pass SequencePosition here (unless there is some perf reason)

Reply:
> You got 2 potential places you could use it in: GetTypeAndIndices (function below)

Yes, I know. I want to keep the change isolated for now. I need to update GetTypeAndIndices to use GetSequenceType as well. 

> Actually I think it would make sense for this function to take int/uint directly and then do that. No reason you need to pass SequencePosition here (unless there is some perf reason)

Good point. Let me play with this.

Reply:
> Actually I think it would make sense for this function to take int/uint directly and then do that. No reason you need to pass SequencePosition here (unless there is some perf reason)

Passing an int instead of SequencePosition has no perf impact. However, now the caller has to pass in position.GetInteger() every time they call GetIndex, which makes the code more noisy. Since there is no harm, I am leaning towards keeping it as is.